{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e9362d8d",
   "metadata": {},
   "source": [
    "# TreeAlign with clone labels as input\n",
    "## Introduction\n",
    "TreeAlign is a model for scRNA and scDNA integration. TreeAlign can take in either total copy number information or allele specific copy number information or both to assign cells from scRNA to a clones identified with scDNA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d600189",
   "metadata": {},
   "source": [
    "## Loading data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6da40fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from treealign import CloneAlignClone\n",
    "from treealign import CloneAlignTree\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import Phylo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "684bb085",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load total copy number input\n",
    "\n",
    "# scRNA read count matrix where each row represents a gene, \n",
    "# each column represents a cell\n",
    "expr = pd.read_csv(\"../data/example_expr.csv\", index_col=0)\n",
    "\n",
    "# scDNA copy number matrix where each row represents a gene,\n",
    "# each column represents a cell\n",
    "# the numbers of the matrix reprents the copy number at given cells and genes\n",
    "cnv = pd.read_csv(\"../data/example_gene_cnv.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8f5f428b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load allele specific input\n",
    "\n",
    "# b allele frequency matrix\n",
    "# each row represents a snp\n",
    "# each column represents a cell\n",
    "# The number in the matrix is the b allele frequency at the given snp and cell\n",
    "hscn = pd.read_csv(\"../data/example_snp_baf.csv\", index_col=0)\n",
    "\n",
    "# reference allele count matrix from scRNA\n",
    "# each row represents a snp\n",
    "# each column represents a cell\n",
    "snv_allele = pd.read_csv(\"../data/example_snp_allele.csv\", index_col=0)\n",
    "\n",
    "# total count matrix at SNPs from scRNA\n",
    "# each row represents a snp\n",
    "# each column represents a cell\n",
    "snv_total = pd.read_csv(\"../data/example_snp_total.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e89638ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# clone labels for each cell in scDNA\n",
    "clone = pd.read_csv(\"../data/example_cell_clone.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "dbe07ea0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_id</th>\n",
       "      <th>clone_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SPECTRUM-OV-022_S1_LEFT_ADNEXA-A108833A-R03-C08</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SPECTRUM-OV-022_S1_LEFT_ADNEXA-A108833A-R03-C09</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SPECTRUM-OV-022_S1_LEFT_ADNEXA-A108833A-R03-C10</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPECTRUM-OV-022_S1_LEFT_ADNEXA-A108833A-R03-C13</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SPECTRUM-OV-022_S1_LEFT_ADNEXA-A108833A-R03-C14</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1057</th>\n",
       "      <td>SPECTRUM-OV-022_S1_RIGHT_ADNEXA-A98179A-R30-C61</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1058</th>\n",
       "      <td>SPECTRUM-OV-022_S1_RIGHT_ADNEXA-A98179A-R30-C64</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1059</th>\n",
       "      <td>SPECTRUM-OV-022_S1_RIGHT_ADNEXA-A98179A-R30-C65</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060</th>\n",
       "      <td>SPECTRUM-OV-022_S1_RIGHT_ADNEXA-A98179A-R30-C67</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1061</th>\n",
       "      <td>SPECTRUM-OV-022_S1_RIGHT_ADNEXA-A98179A-R30-C69</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1062 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              cell_id clone_id\n",
       "0     SPECTRUM-OV-022_S1_LEFT_ADNEXA-A108833A-R03-C08        C\n",
       "1     SPECTRUM-OV-022_S1_LEFT_ADNEXA-A108833A-R03-C09        A\n",
       "2     SPECTRUM-OV-022_S1_LEFT_ADNEXA-A108833A-R03-C10        B\n",
       "3     SPECTRUM-OV-022_S1_LEFT_ADNEXA-A108833A-R03-C13        A\n",
       "4     SPECTRUM-OV-022_S1_LEFT_ADNEXA-A108833A-R03-C14        C\n",
       "...                                               ...      ...\n",
       "1057  SPECTRUM-OV-022_S1_RIGHT_ADNEXA-A98179A-R30-C61        B\n",
       "1058  SPECTRUM-OV-022_S1_RIGHT_ADNEXA-A98179A-R30-C64        C\n",
       "1059  SPECTRUM-OV-022_S1_RIGHT_ADNEXA-A98179A-R30-C65        B\n",
       "1060  SPECTRUM-OV-022_S1_RIGHT_ADNEXA-A98179A-R30-C67        B\n",
       "1061  SPECTRUM-OV-022_S1_RIGHT_ADNEXA-A98179A-R30-C69        C\n",
       "\n",
       "[1062 rows x 2 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# there are four clones in the example datasets: clone A, B, C, None\n",
    "clone"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46c4cd13",
   "metadata": {},
   "source": [
    "## Running TreeAlign with clone labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "915f0468",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene count: 1675\n",
      "cell count: 1000\n",
      "seed = 1\n",
      "Start Inference.\n"
     ]
    }
   ],
   "source": [
    "# construct CloneAlignTree object for data preprocessing\n",
    "\n",
    "# `repeat` is set to 1 here for demonstration purposes. it would be better to set `repeat` larger than 5. \n",
    "# obj = CloneAlignClone(clone=clone, expr=expr, cnv=cnv, hscn=hscn, snv_allele=snv_allele, snv=snv_total, repeat=1)\n",
    "\n",
    "# it is possible to run TreeAlign with total copy number data only\n",
    "obj = CloneAlignClone(clone=clone, expr=expr, cnv=cnv, repeat=1)\n",
    "\n",
    "# it is also possible to run TreeAlign with allele specific data only\n",
    "# obj = CloneAlignClone(clone=clone, hscn=hscn, snv_allele=snv_allele, snv=snv_total, repeat=1)\n",
    "\n",
    "# running TreeAlign to assign cells to phylogenetic subclades\n",
    "obj.assign_cells_to_clones()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "795b4a93",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class CloneAlignClone in module treealign.clonealign_clone:\n",
      "\n",
      "class CloneAlignClone(treealign.clonealign.CloneAlign)\n",
      " |  CloneAlignClone(clone, expr=None, cnv=None, hscn=None, snv_allele=None, snv=None, normalize_cnv=True, cnv_cutoff=10, infer_s_score=True, infer_b_allele=True, repeat=10, min_clone_assign_prob=0.8, min_clone_assign_freq=0.7, min_consensus_gene_freq=0.6, min_consensus_snv_freq=0.6, max_temp=1.0, min_temp=0.5, anneal_rate=0.01, learning_rate=0.1, max_iter=400, rel_tol=5e-05, record_input_output=False, min_clone_cell_count=10)\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      CloneAlignClone\n",
      " |      treealign.clonealign.CloneAlign\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, clone, expr=None, cnv=None, hscn=None, snv_allele=None, snv=None, normalize_cnv=True, cnv_cutoff=10, infer_s_score=True, infer_b_allele=True, repeat=10, min_clone_assign_prob=0.8, min_clone_assign_freq=0.7, min_consensus_gene_freq=0.6, min_consensus_snv_freq=0.6, max_temp=1.0, min_temp=0.5, anneal_rate=0.01, learning_rate=0.1, max_iter=400, rel_tol=5e-05, record_input_output=False, min_clone_cell_count=10)\n",
      " |      initialize CloneAlignClone object\n",
      " |      :param expr: expr read count matrix. row is gene, column is cell. (pandas.DataFrame)\n",
      " |      :param cnv: cnv matrix. row is gene, column is cell. (pandas.DataFrame)\n",
      " |      :param clone: groupings of cnv cells. (pandas.DataFrame)\n",
      " |      :param normalize_cnv: whether to normalized cnv matrix by min or not. (bool)\n",
      " |      :param cnv_cutoff: set cnv higher than cnv_cutoff to cnv_cutoff. (int)\n",
      " |      :param model_select: \"gene\" for the extended clonealign model or \"default\" for the original clonelign model (str)\n",
      " |      :param repeat: num of times to run clonealign to generate consensus results. (int)\n",
      " |      :param min_clone_assign_prob: assign cells to a clone if clone assignment prob reaches min_clone_assign_prob (float)\n",
      " |      :param min_clone_assign_freq: assign cells to a clone if a min proportion of runs generate the same results (float)\n",
      " |      :param max_temp: starting temperature in Gumbel-Softmax reparameterization. (float)\n",
      " |      :param min_temp: min temperature in Gumbel-Softmax reparameterization. (float)\n",
      " |      :param anneal_rate: annealing rate in Gumbel-Softmax reparameterization. (float)\n",
      " |      :param learning_rate: learning rate of Adam optimizer. (float)\n",
      " |      :param max_iter: max number of iterations of elbo optimization during inference. (int)\n",
      " |      :param rel_tol: when the relative change in elbo drops to rel_tol, stop inference. (float)\n",
      " |  \n",
      " |  assign_cells_to_clones(self)\n",
      " |      assign cells from scRNA to clones identified in scDNA\n",
      " |      :return: clone_assign_df (pandas.DataFrame) and gene_type_score_df (pandas.DataFrame)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from treealign.clonealign.CloneAlign:\n",
      " |  \n",
      " |  average_param_dict(self, param_dict)\n",
      " |      # inplace method\n",
      " |  \n",
      " |  clonealign_pyro_model(self, cnv, expr, hscn, snv_allele, snv, infer_s_score=True, infer_b_allele=True, temperature=0.5)\n",
      " |      original clonealign model\n",
      " |      :param cnv: torch.tensor\n",
      " |      :param expr: torch.tensor\n",
      " |      :param hscn:  torch.tensor\n",
      " |      :param snv_allele: torch.tensor\n",
      " |      :param snv: torch.tensor\n",
      " |      :param gene_model: bool\n",
      " |      :param infer_b_allele: bool\n",
      " |      :param temperature: float\n",
      " |      :return: None\n",
      " |  \n",
      " |  construct_allele_specific_input(self, terminals, expr_cells)\n",
      " |  \n",
      " |  construct_total_copy_number_input(self, terminals, expr_cells)\n",
      " |  \n",
      " |  convert_df_to_torch(self, df)\n",
      " |  \n",
      " |  generate_output(self)\n",
      " |      generate clone_assign_df and gene_type_score_df\n",
      " |      :return: clonealign output (pandas.DataFrame)\n",
      " |  \n",
      " |  make_columns_consistent(self, *args)\n",
      " |      # inplace method\n",
      " |  \n",
      " |  max_param_dict(self, param_dict)\n",
      " |      # inplace method\n",
      " |  \n",
      " |  process_input_matrices(self)\n",
      " |      clean up clonealign input\n",
      " |      :return: None\n",
      " |  \n",
      " |  run_clonealign_pyro(self, cnv, expr, hscn, snv_allele, snv, current_repeat)\n",
      " |      clonealign inference\n",
      " |      :param cnv: clone-specific consensus copy number matrix. row is clone. column is gene. (torch.tensor)\n",
      " |      :param expr: gene expression count matrix. row is cell. column is gene. (torch.tensor)\n",
      " |      :return: clone assignment, gene_type_score\n",
      " |  \n",
      " |  run_clonealign_pyro_repeat(self, cnv_df, expr_df, hscn_df, snv_allele_df, snv_df)\n",
      " |      call run_clonealign_pyro() multiple times to generate consensus results\n",
      " |      :param cnv_df: clone-specific consensus copy number matrix. row is gene. column is clone. (pandas.DataFrame)\n",
      " |      :param expr_df: gene expression count matrix. row is gene. column is cell. (pandas.DataFrame)\n",
      " |      :return: frequency of unassigned cells, clone assignment, gene_type_score\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from treealign.clonealign.CloneAlign:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# to view more details about parameters you can customize when you run TreeAlign\n",
    "help(CloneAlignClone)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "012326b5",
   "metadata": {},
   "source": [
    "## Getting results\n",
    "The output of TreeAlign includes: 1. a table indicating the clone to which the cells in scRNA data are assigned. 2. for each gene, a score ranging between 0 and 1 reflecting dosage effects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "50489aaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "clone_assign_df, gene_type_score_df, allele_assign_prob_df = obj.generate_output()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7264571b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_id</th>\n",
       "      <th>clone_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_CTAGACAT...</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_CTCCATGA...</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_TTGGGTAG...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_CCCTCAAC...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_TTGTTTGA...</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_AGCTACAT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_LEFT_ADNEXA_TCCTGCAGT...</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_GTTACCCC...</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_LEFT_ADNEXA_ATTTCACTC...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_ATTGTTCT...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               cell_id clone_id\n",
       "0    SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_CTAGACAT...        C\n",
       "1    SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_CTCCATGA...        B\n",
       "2    SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_TTGGGTAG...     None\n",
       "3    SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_CCCTCAAC...     None\n",
       "4    SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_TTGTTTGA...        A\n",
       "..                                                 ...      ...\n",
       "995  SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_AGCTACAT...     None\n",
       "996  SPECTRUM-OV-022_S1_CD45N_LEFT_ADNEXA_TCCTGCAGT...        C\n",
       "997  SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_GTTACCCC...        C\n",
       "998  SPECTRUM-OV-022_S1_CD45N_LEFT_ADNEXA_ATTTCACTC...     None\n",
       "999  SPECTRUM-OV-022_S1_CD45N_RIGHT_ADNEXA_ATTGTTCT...     None\n",
       "\n",
       "[1000 rows x 2 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subclade assignment for each cell in scRNA data\n",
    "clone_assign_df\n",
    "\n",
    "# save clone assignment results to csv file\n",
    "# clone_assign_df.to_csv(\"test.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "95871f13",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_type_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TMEM53</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KIF2C</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RPS8</td>\n",
       "      <td>9.943840e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BEST4</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PLK3</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1670</th>\n",
       "      <td>ATP2B3</td>\n",
       "      <td>6.474631e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1671</th>\n",
       "      <td>DUSP9</td>\n",
       "      <td>1.211680e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1672</th>\n",
       "      <td>PNCK</td>\n",
       "      <td>9.999979e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1673</th>\n",
       "      <td>SLC6A8</td>\n",
       "      <td>9.999830e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1674</th>\n",
       "      <td>BCAP31</td>\n",
       "      <td>9.103891e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1675 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene  gene_type_score\n",
       "0     TMEM53     1.000000e+00\n",
       "1      KIF2C     1.000000e+00\n",
       "2       RPS8     9.943840e-01\n",
       "3      BEST4     1.000000e+00\n",
       "4       PLK3     1.000000e+00\n",
       "...      ...              ...\n",
       "1670  ATP2B3     6.474631e-03\n",
       "1671   DUSP9     1.211680e-11\n",
       "1672    PNCK     9.999979e-01\n",
       "1673  SLC6A8     9.999830e-01\n",
       "1674  BCAP31     9.103891e-01\n",
       "\n",
       "[1675 rows x 2 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# the probability of having dosage effects for each gene\n",
    "gene_type_score_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "341831f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele_assign_prob_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5548ad0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
