{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2dc7878b",
   "metadata": {},
   "source": [
    "# TreeAlign with clone labels as input\n",
    "## Introduction\n",
    "TreeAlign can also take in 1. a cell by gene `scRNA count matrix`, 2. cell by gene `scDNA copy number matrix` and 3. a `clone label` table which indicates the clone label for each cell in scDNA. TreeAlign assigns cells from the scRNA count matrix to a clone in scDNA. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed258264",
   "metadata": {},
   "source": [
    "## Loading data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0c3dc6ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "from treealign import CloneAlignClone\n",
    "from treealign import CloneAlignTree\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import Phylo\n",
    "\n",
    "# scRNA read count matrix where each row represents a gene, \n",
    "# each column represents a cell\n",
    "expr = pd.read_csv(\"../data/example_expr.csv\", index_col=0)\n",
    "\n",
    "# scDNA copy number matrix where each row represents a gene,\n",
    "# each column represents a cell\n",
    "cnv = pd.read_csv(\"../data/example_gene_cnv.csv\", index_col=0)\n",
    "\n",
    "# clone labels for each cell in scDNA\n",
    "clone = pd.read_csv(\"../data/example_cell_clone.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "68c8aa82",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_id</th>\n",
       "      <th>clone_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>081LA_DLP_UNSORTED-128673A-R14-C25</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>081LA_DLP_UNSORTED-128673A-R14-C44</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>081LA_DLP_UNSORTED-128673A-R14-C46</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>081LA_DLP_UNSORTED-128673A-R14-C60</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>081LA_DLP_UNSORTED-128673A-R15-C12</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>081LA_DLP_UNSORTED-128673A-R32-C49</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>081LA_DLP_UNSORTED-128673A-R32-C50</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>081LA_DLP_UNSORTED-128673A-R32-C54</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>081LA_DLP_UNSORTED-128673A-R32-C56</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>081LA_DLP_UNSORTED-128673A-R32-C57</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>180 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                cell_id clone_id\n",
       "0    081LA_DLP_UNSORTED-128673A-R14-C25        C\n",
       "1    081LA_DLP_UNSORTED-128673A-R14-C44        A\n",
       "2    081LA_DLP_UNSORTED-128673A-R14-C46        A\n",
       "3    081LA_DLP_UNSORTED-128673A-R14-C60        B\n",
       "4    081LA_DLP_UNSORTED-128673A-R15-C12        C\n",
       "..                                  ...      ...\n",
       "175  081LA_DLP_UNSORTED-128673A-R32-C49        C\n",
       "176  081LA_DLP_UNSORTED-128673A-R32-C50        A\n",
       "177  081LA_DLP_UNSORTED-128673A-R32-C54        A\n",
       "178  081LA_DLP_UNSORTED-128673A-R32-C56        C\n",
       "179  081LA_DLP_UNSORTED-128673A-R32-C57        C\n",
       "\n",
       "[180 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clone"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa9268c7",
   "metadata": {},
   "source": [
    "## Running TreeAlign with clone labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d87b2558",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start run clonealign for 3 clones:\n",
      "cnv gene count: 498\n",
      "expr cell count: 1000\n",
      "Start Inference.\n",
      "\n",
      "..................................................................................................................................................ELBO converged at iteration 147\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(    cell_id clone_id\n",
       " 0         0        C\n",
       " 1         1        C\n",
       " 2         2        A\n",
       " 3         3        C\n",
       " 4         4        B\n",
       " ..      ...      ...\n",
       " 995     995        A\n",
       " 996     996        B\n",
       " 997     997        B\n",
       " 998     998        A\n",
       " 999     999        B\n",
       " \n",
       " [1000 rows x 2 columns],\n",
       "          gene  gene_type_score\n",
       " 0      NMNAT1     9.961327e-01\n",
       " 1    C1orf127     4.503274e-10\n",
       " 2      MAD2L2     9.990683e-01\n",
       " 3    KIAA2013     9.865901e-01\n",
       " 4       PLOD1     9.247180e-01\n",
       " ..        ...              ...\n",
       " 493      ASPN     8.546441e-03\n",
       " 494     CENPP     9.996878e-01\n",
       " 495      ECM2     6.237036e-08\n",
       " 496        GK     1.954867e-02\n",
       " 497      TAB3     9.925644e-01\n",
       " \n",
       " [498 rows x 2 columns],\n",
       "      0\n",
       " 0    C\n",
       " 1    C\n",
       " 2    A\n",
       " 3    C\n",
       " 4    B\n",
       " ..  ..\n",
       " 995  A\n",
       " 996  B\n",
       " 997  B\n",
       " 998  A\n",
       " 999  B\n",
       " \n",
       " [1000 rows x 1 columns],\n",
       "                      0\n",
       " NMNAT1    9.961327e-01\n",
       " C1orf127  4.503274e-10\n",
       " MAD2L2    9.990683e-01\n",
       " KIAA2013  9.865901e-01\n",
       " PLOD1     9.247180e-01\n",
       " ...                ...\n",
       " ASPN      8.546441e-03\n",
       " CENPP     9.996878e-01\n",
       " ECM2      6.237036e-08\n",
       " GK        1.954867e-02\n",
       " TAB3      9.925644e-01\n",
       " \n",
       " [498 rows x 1 columns])"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# construct CloneAlignTree object for data preprocessing\n",
    "obj = CloneAlignClone(expr, cnv, clone, repeat=1)\n",
    "\n",
    "# running TreeAlign to assign cells to phylogenetic subclades\n",
    "obj.assign_cells_to_clones()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c5ce575b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class CloneAlignClone in module treealign.clonealign_clone:\n",
      "\n",
      "class CloneAlignClone(treealign.clonealign.CloneAlign)\n",
      " |  Method resolution order:\n",
      " |      CloneAlignClone\n",
      " |      treealign.clonealign.CloneAlign\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, expr, cnv, clone, normalize_cnv=True, cnv_cutoff=10, model_select='gene', repeat=10, min_clone_cell_count=20, min_clone_assign_prob=0.8, min_clone_assign_freq=0.7, min_consensus_gene_freq=0.2, max_temp=1.0, min_temp=0.5, anneal_rate=0.01, learning_rate=0.1, max_iter=400, rel_tol=5e-05)\n",
      " |      initialize CloneAlignClone object\n",
      " |      :param expr: expr read count matrix. row is gene, column is cell. (pandas.DataFrame)\n",
      " |      :param cnv: cnv matrix. row is gene, column is cell. (pandas.DataFrame)\n",
      " |      :param clone: groupings of cnv cells. (pandas.DataFrame)\n",
      " |      :param normalize_cnv: whether to normalized cnv matrix by min or not. (bool)\n",
      " |      :param cnv_cutoff: set cnv higher than cnv_cutoff to cnv_cutoff. (int)\n",
      " |      :param model_select: \"gene\" for the extended clonealign model or \"default\" for the original clonelign model (str)\n",
      " |      :param repeat: num of times to run clonealign to generate consensus results. (int)\n",
      " |      :param min_clone_assign_prob: assign cells to a clone if clone assignment prob reaches min_clone_assign_prob (float)\n",
      " |      :param min_clone_assign_freq: assign cells to a clone if a min proportion of runs generate the same results (float)\n",
      " |      :param max_temp: starting temperature in Gumbel-Softmax reparameterization. (float)\n",
      " |      :param min_temp: min temperature in Gumbel-Softmax reparameterization. (float)\n",
      " |      :param anneal_rate: annealing rate in Gumbel-Softmax reparameterization. (float)\n",
      " |      :param learning_rate: learning rate of Adam optimizer. (float)\n",
      " |      :param max_iter: max number of iterations of elbo optimization during inference. (int)\n",
      " |      :param rel_tol: when the relative change in elbo drops to rel_tol, stop inference. (float)\n",
      " |  \n",
      " |  assign_cells_to_clones(self)\n",
      " |      assign cells from scRNA to clones identified in scDNA\n",
      " |      :return: clone_assign_df (pandas.DataFrame) and gene_type_score_df (pandas.DataFrame)\n",
      " |  \n",
      " |  generate_output(self)\n",
      " |      generate clone_assign_df and gene_type_score_df\n",
      " |      :return: clonealign output (pandas.DataFrame)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from treealign.clonealign.CloneAlign:\n",
      " |  \n",
      " |  clonealign_pyro_gene_model(self, cnv, expr, temperature=0.5)\n",
      " |      extended clonealign model\n",
      " |      :param cnv: torch.tensor\n",
      " |      :param expr: torch.tensor\n",
      " |      :param temperature: float\n",
      " |      :return: None\n",
      " |  \n",
      " |  clonealign_pyro_model(self, cnv, expr, temperature=0.5)\n",
      " |      original clonealign model\n",
      " |      :param cnv: torch.tensor\n",
      " |      :param expr: torch.tensor\n",
      " |      :param temperature: float\n",
      " |      :return: None\n",
      " |  \n",
      " |  process_input_matrices(self)\n",
      " |      clean up clonealign input\n",
      " |      :return: None\n",
      " |  \n",
      " |  run_clonealign_pyro(self, cnv, expr)\n",
      " |      clonealign inference\n",
      " |      :param cnv: clone-specific consensus copy number matrix. row is clone. column is gene. (torch.tensor)\n",
      " |      :param expr: gene expression count matrix. row is cell. column is gene. (torch.tensor)\n",
      " |      :return: clone assignment, gene_type_score\n",
      " |  \n",
      " |  run_clonealign_pyro_repeat(self, cnv_df, expr_df)\n",
      " |      call run_clonealign_pyro() multiple times to generate consensus results\n",
      " |      :param cnv_df: clone-specific consensus copy number matrix. row is gene. column is clone. (pandas.DataFrame)\n",
      " |      :param expr_df: gene expression count matrix. row is gene. column is cell. (pandas.DataFrame)\n",
      " |      :return: frequency of unassigned cells, clone assignment, gene_type_score\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from treealign.clonealign.CloneAlign:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# to view more details about parameters you can customize when you run TreeAlign\n",
    "help(CloneAlignClone)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50c555b3",
   "metadata": {},
   "source": [
    "## Getting results\n",
    "The output of TreeAlign includes: 1. a table indicating the clone to which the cells in scRNA data are assigned. 2. for each gene, a score ranging between 0 and 1 reflecting dosage effects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0c4bcab8",
   "metadata": {},
   "outputs": [],
   "source": [
    "clone_assign, s_score, clone_assign_raw, s_score_raw = obj.generate_output()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "90fa2214",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_id</th>\n",
       "      <th>clone_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>995</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>996</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>997</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>998</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>999</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    cell_id clone_id\n",
       "0         0        C\n",
       "1         1        C\n",
       "2         2        A\n",
       "3         3        C\n",
       "4         4        B\n",
       "..      ...      ...\n",
       "995     995        A\n",
       "996     996        B\n",
       "997     997        B\n",
       "998     998        A\n",
       "999     999        B\n",
       "\n",
       "[1000 rows x 2 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subclade assignment for each cell in scRNA data\n",
    "clone_assign"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b94bfe1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_type_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NMNAT1</td>\n",
       "      <td>9.961327e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C1orf127</td>\n",
       "      <td>4.503274e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MAD2L2</td>\n",
       "      <td>9.990683e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KIAA2013</td>\n",
       "      <td>9.865901e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PLOD1</td>\n",
       "      <td>9.247180e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493</th>\n",
       "      <td>ASPN</td>\n",
       "      <td>8.546441e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>494</th>\n",
       "      <td>CENPP</td>\n",
       "      <td>9.996878e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495</th>\n",
       "      <td>ECM2</td>\n",
       "      <td>6.237036e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496</th>\n",
       "      <td>GK</td>\n",
       "      <td>1.954867e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>TAB3</td>\n",
       "      <td>9.925644e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>498 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene  gene_type_score\n",
       "0      NMNAT1     9.961327e-01\n",
       "1    C1orf127     4.503274e-10\n",
       "2      MAD2L2     9.990683e-01\n",
       "3    KIAA2013     9.865901e-01\n",
       "4       PLOD1     9.247180e-01\n",
       "..        ...              ...\n",
       "493      ASPN     8.546441e-03\n",
       "494     CENPP     9.996878e-01\n",
       "495      ECM2     6.237036e-08\n",
       "496        GK     1.954867e-02\n",
       "497      TAB3     9.925644e-01\n",
       "\n",
       "[498 rows x 2 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# the probability of having dosage effects for each gene\n",
    "s_score"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
